[aotinductor] Fix duplicated unbacked symbol declarations

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111973
* #112021
* __->__ #111823

Summary: For https://github.com/pytorch/pytorch/issues/111711

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

desertfire committed:
[aotinductor] Fix duplicated unbacked symbol declarations

Summary: For https://github.com/pytorch/pytorch/issues/111711

[ghstack-poisoned]

chenyang78 commented on torch/_inductor/ir.py:
```diff
@@ -2694,18 +2694,18 @@ def codegen_unbacked_symbol_defs(self, wrapper):
             if s in symbols_to_define:
                 wrapper.writeline(
-                    f"{wrapper.declare}{s} = {self.get_name()}.size({i}){wrapper.ending}"
+                    f"{wrapper.codegen_unbacked_symbol_decl(s)} = {self.get_name()}.size({i}){wrapper.ending}"
```
A naive question - where do the duplicate unbacked symbols come from in the first place? I am not familiar with how the unbacked symbol defs are created, so I am wondering if it is legitimate to have such dups?

desertfire replied:
Good question. In the test example, it looks like the second `index.Tensor` should be CSE-ed. I think unbacked symbols should be unique in general (can't reuse symbols just for any random expressions). @ezyang would be the best person to answer this question.

```
                index_1 = torch.ops.aten.index.Tensor(x, [eq])
                sin = torch.ops.aten.sin.default(index_1)
                index_2 = torch.ops.aten.index.Tensor(x, [eq])
```

desertfire commented:
@pytorchbot merge

ezyang reviewed (approved):
Thanks!

desertfire committed:
Update on "[aotinductor] Fix duplicated unbacked symbol declarations"


Summary: For https://github.com/pytorch/pytorch/issues/111711

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire committed:
Update on "[aotinductor] Fix duplicated unbacked symbol declarations"


Summary: For https://github.com/pytorch/pytorch/issues/111711

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

desertfire committed:
Update on "[aotinductor] Fix duplicated unbacked symbol declarations"


Summary: For https://github.com/pytorch/pytorch/issues/111711

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]
